plugin/trino-elasticsearch/src/main/java/io/trino/plugin/elasticsearch/AggregationQueryPageSource.java

```diff
@@ -0,0 +1,211 @@
+        List<Map<String, Object>> flatResult = getResult(searchResponse);
+        after = extractAfter(searchResponse);
+        for (Map<String, Object> result : flatResult) {
+            for (int i = 0; i < columns.size(); i++) {
+                String key = columns.get(i).getName();
+                decoders.get(i).decode(noUsed, () -> result.get(key), columnBuilders[i]);
+            }
+        }
+        Block[] blocks = new Block[columnBuilders.length];
+        for (int i = 0; i < columnBuilders.length; i++) {
+            blocks[i] = columnBuilders[i].build();
+            columnBuilders[i] = columnBuilders[i].newBlockBuilderLike(null);
+        }
+        return new Page(blocks);
+    }
+
+    // TODO consider support 6.1/6.2
+    //  6.1/6.2: no after_key, just use last bucket as the next retrieve parameter
+    //  >= 6.3: the result contains after_key, use after_key as the next retrieve
+    private Optional<Map<String, Object>> extractAfter(SearchResponse searchResponse)
+    {
+        Aggregations aggregations = searchResponse.getAggregations();
+        if (aggregations == null) {
+            return Optional.empty();
+        }
+        for (Aggregation agg : aggregations) {
+            if (agg instanceof CompositeAggregation) {
+                CompositeAggregation compositeAggregation = (CompositeAggregation) agg;
+                return Optional.ofNullable(compositeAggregation.afterKey());
+            }
+        }
+        return Optional.empty();
+    }
+
+    private List<Map<String, Object>> getResult(SearchResponse searchResponse)
+    {
+        Aggregations aggregations = searchResponse.getAggregations();
+        if (aggregations == null) {
+            return Collections.emptyList();
+        }
+        Map<String, Object> singleValueMap = new LinkedHashMap<>();
+        List<Map<String, Object>> result = new ArrayList<>();
+        boolean hasBucketsAggregation = false;
+        for (Aggregation agg : aggregations) {
+            if (agg instanceof CompositeAggregation) {
+                hasBucketsAggregation = true;
+                CompositeAggregation compositeAggregation = (CompositeAggregation) agg;
+                for (CompositeAggregation.Bucket bucket : compositeAggregation.getBuckets()) {
+                    Map<String, Object> line = new HashMap<>();
+                    line.putAll(bucket.getKey());
+                    for (Aggregation metricAgg : bucket.getAggregations()) {
+                        if (metricAgg instanceof NumericMetricsAggregation.SingleValue) {
+                            NumericMetricsAggregation.SingleValue sv = (NumericMetricsAggregation.SingleValue) metricAgg;
+                            line.put(metricAgg.getName(), sv.value());
+                        }
+                    }
+                    result.add(line);
+                }
+            }
+            else if (agg instanceof NumericMetricsAggregation.SingleValue) {
+                if (hasBucketsAggregation) {
```

Comment:
This check relies on the assumption that when iterating over aggregations, the composite would be encountered first.